### PR TITLE
chore(feat): localize 404 page (fix #1987)

### LINKED ISSUE #1987: 404 page does not support i18n
- [x] I confirm that this is an issue rather than a question.

## Bug report

I just find this on [stackoverflow](https://stackoverflow.com/questions/58579031/is-translation-supposed-to-be-working-for-notfound-vue), and test it on the official website

#### Steps to reproduce

Try to visit [vuepress.vuejs.org/zh/test](https://vuepress.vuejs.org/zh/test)

#### What is expected?

A 404 page with Chinese and

```html
<html lang="zh-CN" >
 ```

#### What is actually happening?

```html
<html lang="en-US" >
 ```

#### Other relevant information

- Output of `npx vuepress info` in my VuePress project:

### PR DESCRIPTION
**Pull request by dzetah**
**Summary**

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

This PR adds localized messages to the 404 page, matching the language of the URL.
It is still missing some Chinese translations for the default messages which I am not able to provide.

### CODE REVIEW

File: packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js [hunk 1]
```diff
@@ -21,13 +21,27 @@ module.exports = {
...
         selectText: '选择语言',
         editLinkText: '在 GitHub 上编辑此页',
         lastUpdated: '上次更新',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
+          `There's nothing here.`,
+          `How did we get here?`,
+          `That's a Four-Oh-Four.`,
+          `Looks like we've got some broken links.`
+        ]
```
Comment (dzetah) [inline packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js]: Needs translation.

File: packages/@vuepress/test-utils/client/siteData.js [hunk 2]
```diff
@@ -29,13 +29,27 @@ export default {
...
         "editLinkText": "在 GitHub 上编辑此页",
-        "lastUpdated": "上次更新"
+        "lastUpdated": "上次更新",
+        "notFoundLinkText": "Take me home.",
+        "notFoundMessages": [
+          "There's nothing here.",
+          "How did we get here?",
+          "That's a Four-Oh-Four.",
+          "Looks like we've got some broken links."
+        ]
```
Comment (dzetah) [inline packages/@vuepress/test-utils/client/siteData.js]: Needs translation.

File: packages/docs/docs/.vuepress/config.js [hunk 3]
```diff
@@ -56,6 +63,13 @@ module.exports = ctx => ({
         selectText: '选择语言',
         ariaLabel: "选择语言",
         editLinkText: '在 GitHub 上编辑此页',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
+          `There's nothing here.`,
+          `How did we get here?`,
+          `That's a Four-Oh-Four.`,
+          `Looks like we've got some broken links.`
+        ],
```
Comment (dzetah) [inline packages/docs/docs/.vuepress/config.js]: Needs translation.

File: packages/@vuepress/core/lib/node/ClientComputedMixin.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ module.exports = siteData => {
       for (const path in locales) {
         if (path === '/') {
           defaultLang = locales[path]
-        } else if (this.$page.path.indexOf(path) === 0) {
+        } else if (this.$page.path.indexOf(path) === 0 || (!this.$page.path && this.$route.path.indexOf(path) === 0)) {
```
Comment (kefranabg) [inline packages/@vuepress/core/lib/node/ClientComputedMixin.js]: What are you trying to solve with this condition? When would `this.$page.path` not be defined?

File: packages/docs/docs/.vuepress/config.js [hunk 5]
```diff
@@ -42,6 +42,13 @@ module.exports = ctx => ({
         selectText: 'Languages',
         ariaLabel: 'Select language',
         editLinkText: 'Edit this page on GitHub',
+        notFoundLinkText: 'Take me home.',
```
Comment (kefranabg) [inline packages/docs/docs/.vuepress/config.js]: I don't think we need to set this for the `/` locale as these are the default values right?

File: packages/docs/docs/zh/guide/i18n.md [hunk 6]
```diff
@@ -76,6 +78,10 @@ module.exports = {
         label: '简体中文',
         // 编辑链接文字
         editLinkText: '在 GitHub 上编辑此页',
+        // TODO: comment me
+        notFoundLinkText: 'Take me home.',
+        // TODO: comment me
+        notFoundMessages: [/* ... */],
```
Comment (kefranabg) [inline packages/docs/docs/zh/guide/i18n.md]: Same here (to translate)

File: packages/docs/docs/zh/guide/i18n.md [hunk 7]
```diff
@@ -54,6 +54,8 @@ module.exports = {
         label: 'English',
         ariaLabel: 'Languages'
         editLinkText: 'Edit this page on GitHub',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [/* ... */],
```
Comment (kefranabg) [inline packages/docs/docs/zh/guide/i18n.md]: Same here (to translate)

File: packages/docs/docs/theme/default-theme-config.md [hunk 8]
```diff
@@ -454,7 +454,13 @@ module.exports = {
...
     prevLinks: false,
     // custom text for edit link. Defaults to "Edit this page"
-    editLinkText: 'Help us improve this page!'
+    editLinkText: 'Help us improve this page!',
+
+    // Optional options fot the 404 page
+
+    // custom text for 404 page link. Defaults to "Take me home."
+    notFoundLinkText: 'Take me home.',
+    notFoundMessages: [/* ... */]
```
Comment (kefranabg) [inline packages/docs/docs/theme/default-theme-config.md]: I think this is important to write an example with the default values here, to let the user know what the default values are.

File: packages/docs/docs/guide/i18n.md [hunk 9]
```diff
@@ -79,6 +83,8 @@ module.exports = {
         selectText: '选择语言',
         label: '简体中文',
         editLinkText: '在 GitHub 上编辑此页',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [/* ... */],
```
Comment (kefranabg) [inline packages/docs/docs/guide/i18n.md]: Same thing here

File: packages/docs/docs/guide/i18n.md [hunk 10]
```diff
@@ -58,6 +58,10 @@ module.exports = {
         ariaLabel: 'Languages'
         // text for the edit-on-github link
         editLinkText: 'Edit this page on GitHub',
+        // text for the 404 page link
+        notFoundLinkText: 'Take me home.',
+        // the list of random messages to show on the 404 page
+        notFoundMessages: [/* ... */],
```
Comment (kefranabg) [inline packages/docs/docs/guide/i18n.md]: I think this is important to write an example with the default values here, to let the user know what the default values are.

File: packages/docs/docs/.vuepress/config.js [hunk 11]
```diff
@@ -56,6 +63,13 @@ module.exports = ctx => ({
         selectText: '选择语言',
         ariaLabel: "选择语言",
         editLinkText: '在 GitHub 上编辑此页',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
```
Comment (kefranabg) [inline packages/docs/docs/.vuepress/config.js]: Use simple quotes here

File: packages/docs/docs/.vuepress/config.js [hunk 12]
```diff
@@ -42,6 +42,13 @@ module.exports = ctx => ({
         selectText: 'Languages',
         ariaLabel: 'Select language',
         editLinkText: 'Edit this page on GitHub',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
```
Comment (kefranabg) [inline packages/docs/docs/.vuepress/config.js]: Use simple quotes here

File: packages/@vuepress/theme-default/layouts/404.vue [hunk 13]
```diff
@@ -3,23 +3,42 @@
...
+    },
+    notFoundLinkText () {
+      return (
+        this.$themeLocaleConfig.notFoundLinkText
+        || this.$site.themeConfig.notFoundLinkText
+        || `Take me home.`
+      )
+    },
+    notFoundMessages () {
+      const messages = [
```
Comment (kefranabg) [inline packages/@vuepress/theme-default/layouts/404.vue]: I know that back quotes were already used before, but here you can use simple quotes 😉

File: packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js [hunk 14]
```diff
@@ -21,13 +21,27 @@ module.exports = {
         label: 'English',
         selectText: 'Languages',
         editLinkText: 'Edit this page on GitHub',
-        lastUpdated: 'Last Updated'
+        lastUpdated: 'Last Updated',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
```
Comment (kefranabg) [inline packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js]: Use simple quotes here

File: packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js [hunk 15]
```diff
@@ -21,13 +21,27 @@ module.exports = {
...
+          `Looks like we've got some broken links.`
+        ]
       },
       '/zh/': {
         label: '简体中文',
         selectText: '选择语言',
         editLinkText: '在 GitHub 上编辑此页',
         lastUpdated: '上次更新',
+        notFoundLinkText: 'Take me home.',
+        notFoundMessages: [
```
Comment (kefranabg) [inline packages/@vuepress/core/lib/node/__tests__/prepare/fixtures/docs-i18n/.vuepress/config.js]: Use simple quotes here

**Comment by kefranabg** [comment]
@newsbielt703 or @meteorlxy could you help with the `zh` translation?

**Comment by kefranabg** [comment]
@dzetah Sorry I did not that some sentences were using apostrophes 😅 In this case you can use wrap these sentences with `"` instead of using `\'` in the sentences.

**Comment by dzetah** [comment]
Hello, any news on this ? I'd like to see this merged soon :)

I think it only lacks the Chinese translation (docs and defaults).

File: packages/@vuepress/core/lib/node/ClientComputedMixin.js (hunk 4 above)
Comment (kefranabg) [reply packages/@vuepress/core/lib/node/ClientComputedMixin.js]: @dzetah ?

**Comment by kefranabg** [comment]
@dzetah yes, we just need one more approve and zh translation (@meteorlxy ?) BTW could you answer this question in my review ? 

> What are you trying to solve with this condition? When would this.$page.path not be defined?

**Comment by dzetah** [comment]
@kefranabg I answered this question before but it looks like you can't see my answer so i'll post it here below too :)

> When the page wasn't found (before it redirects to the 404 page, `this.$page.path` is set to `''`. So in this case the idea was to match the locale directly from the URL instead of the page path.

**Comment by meteorlxy** [comment]
I've resolved conflicts, linted the code, and made some tweaks.

**Comment by meteorlxy** [comment]
Currently, it will show the default messages first, and change to the locale messages.

It's not so perfect 🤔 

https://deploy-preview-2004--vuepress.netlify.com/zh/guide/aaa

demo-locale

**Comment by kefranabg** [comment]
@dzetah Could you fix this? We're not supposed to have this blink effect

**Comment by meteorlxy** [comment]
Although it's not a critical problem, I think it would be a little complicated to fix that.

Maybe [vuepress-plugin-dehydrate](https://vuepress.github.io/plugins/dehydrate/) will help.

**Comment by dzetah** [comment]
@kefranabg Obviously this is not an expected behaviour, I'll try to take a look at this later this week...

Thanks @meteorlxy for your improvements :)

**Comment by dzetah** [comment]
@meteorlxy I'm not sure [vuepress-plugin-dehydrate](https://vuepress.github.io/plugins/dehydrate/) will help us in this case because the 404 component is replacing the page at the given location. Maybe I'm misunderstanding the issue here 🤔

**Comment by dzetah** [comment]
@meteorlxy @kefranabg I think we could just use the `<client-only>` component for the 404 page content, so it will be ignored during SSR, what do you think ?

**Comment by dzetah** [comment]
Hey @kefranabg is it possible to merge this ?

**Comment by dzetah** [comment]
I opened this PR a year ago and I just stumbled upon it again today 😅

Is this still something that can be merged ?

**Comment by meteorlxy** [comment]
@dzetah See #2744.

BTW, it's already supported in the default theme of VuePress v2

**Comment by dzetah** [comment]
Hi @meteorlxy thanks for your answer, I'm going to use VuePress 2 from now on then :)

I guess we can decline this PR.

**Pull request closed without merging**